This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [1]:
# get api key from your .env file
import os
from dotenv import load_dotenv

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')

#print(API_KEY)

Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [3]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url = 'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X.json?limit=1&api_key=' + API_KEY
r = requests.get(url)

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json_data = r.json()

# Returned a dict with one key 'dataset'
print(json_data.keys())

# Within 'dataset', there are 21 keys
print(json_data['dataset'].keys())

# There is more general information about AFX_X from the FSE database, such as the description and id
# Sample of one day can be seen in 'data', and as of the time of writing this, the date is shown as 2020-12-01
for k in json_data['dataset'].keys():
    print(k + ": " + str(json_data['dataset'][k]))

dict_keys(['dataset'])
dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])
id: 10095370
dataset_code: AFX_X
database_code: FSE
name: Carl Zeiss Meditec (AFX_X)
description: Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704
refreshed_at: 2020-12-01T14:48:09.907Z
newest_available_date: 2020-12-01
oldest_available_date: 2000-06-07
column_names: ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']
frequency: daily
type: Time Series
premium: False
limit: 1
transform: None
column_index: None
start_date: 2000-06-07
end_date: 2020-12-01
data: [['2020-12-01', 112.2, 112.2, 111.

## Mini Project Tasks

### Instructions

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

### Response

#### 1. Collect data from Frankfurst Stock Exchange for AFX_X for year 2017

In [5]:
# 1. Collect data from Frankfurst Stock Exchange for AFX_X for year 2017
mini_url = 'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=' + API_KEY
mini_r = requests.get(mini_url)

#### 2. Convert JSON object into a Python dict

In [6]:
# 2. Returned data saved as dict `mini_json`
mini_json = (mini_r.json())
type(mini_json)

dict

In [7]:
# Peeking the json
for k in mini_json['dataset'].keys():
    if k == 'data':
        print(k + ': ' + str(type(mini_json['dataset'][k])))
    else:
        print(k + ': ' + str(mini_json['dataset'][k]))

id: 10095370
dataset_code: AFX_X
database_code: FSE
name: Carl Zeiss Meditec (AFX_X)
description: Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704
refreshed_at: 2020-12-01T14:48:09.907Z
newest_available_date: 2020-12-01
oldest_available_date: 2000-06-07
column_names: ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']
frequency: daily
type: Time Series
premium: False
limit: None
transform: None
column_index: None
start_date: 2017-01-01
end_date: 2017-12-31
data: <class 'list'>
collapse: None
order: None
database_id: 6129


In [8]:
# list `data_list` containing dict of 'column_names':'data'
col_names = mini_json['dataset']['column_names']
data_list = []
for e in mini_json['dataset']['data']:
    data_list.append(dict(zip(col_names,e)))


#### 3. Calculate the highest and lowest opening prices in 2017

In [9]:
def highlo_price(data_list, col_name):
    """
    This function returns the highest price, lowest price, and possibly excluded dates, of a given column.
    It accepts a list of dictionaries, data_list, which contains the keys col_name and 'Date'.
    
    data_list is a list of dictionaries, in which the keys col_name and 'Date' can be found.
    col_name is a string corresponding to the key to search for the target highest and lowest values.
    """
    # Other errors to check for could include index and length issues
    # Variable instantiation
    highest_price = 0
    lowest_price = None
    i = 0
    excluded_dates = []
    while lowest_price is None and i < len(data_list):
        lowest_price = data_list[i][col_name]
        i = i + 1
    
    # Comparing values in data
    for e in data_list:
        open_price = e[col_name]
        try:
            if open_price > highest_price:
                highest_price = open_price
            if open_price < lowest_price:
                lowest_price = open_price
        except TypeError:
            excluded_dates.append((e['Date'],str(e[col_name])))
    
    return (highest_price, lowest_price, excluded_dates)

In [10]:
# 3. Highest and Lowest opening prices in 2017
highest_price, lowest_price, excluded_dates = highlo_price(data_list, 'Open')

print("Highest Opening Price: %.2f" % highest_price)
print("Lowest Opening Price: %.2f" % lowest_price)
print("Dates excluded: " + str(excluded_dates))

Highest Opening Price: 53.11
Lowest Opening Price: 34.00
Dates excluded: [('2017-05-01', 'None'), ('2017-04-17', 'None'), ('2017-04-14', 'None')]


Highest and lowest opening prices of 2017 were 53.11 and 34.00 respectively.

#### 4. Find the greatest change in price in single day (high and low price)

In [11]:
# 4. Greatest change in single day (high and low price)
greatest_diff = 0

for e in data_list:
    day_diff = e['High'] - e['Low']
    if day_diff > greatest_diff:
        greatest_diff = day_diff

print("Greatest change in a single day: %.2f" % greatest_diff)

Greatest change in a single day: 2.81


The largest change in any one day based on the High and Low prices was 2.81.

#### 5. Find the greatest change in closing price between any two days

In [12]:
# 5. Greatest change between any 2 days using closing price
highest_close, lowest_close, excluded_dates = highlo_price(data_list, 'Close')
greatest_close_change = highest_close - lowest_close
print("Greatest Change in Closing Price between any 2 days: %.2f" % greatest_close_change)

Greatest Change in Closing Price between any 2 days: 19.03


The largest change between any two days based on Closing Price is 19.03.

#### 6. Find the average daily trading volume in 2017

ADTV for n Days = (Total Volume Traded During n Days) / n

In [13]:
# 6. Average daily trading volume in 2017
total_vol_traded = 0
days = 0
for e in data_list:
    total_vol_traded += e['Traded Volume']
    days += 1

adtv = total_vol_traded / days
print("Average Daily Trading Volume in 2017: ADTV = %.2f" % adtv)

Average Daily Trading Volume in 2017: ADTV = 89124.34


The average daily trading volume in 2017 was 89,124.34.

#### 7. (Optional) Find the median daily trading volume in 2017

In [14]:
# 7. (Optional) Median daily trading volume in 2017
tv_list = []
for e in data_list:
    tv_list.append(e['Traded Volume'])

tv_list.sort()
#print(tv_list[int(len(tv_list)/2)])
#print(tv_list[int(len(tv_list)/2) + 1])
tv_med = (tv_list[int(len(tv_list)/2)] + tv_list[int(len(tv_list)/2) + 1]) / 2.0
print("Median Daily Trading Volume in 2017: %.2f" % tv_med)

Median Daily Trading Volume in 2017: 76443.00


The median trading volume during this year was 76,443.00.